pytensor/link/jax/dispatch/scan.py

```diff
@@ -1,159 +1,199 @@
-            inner_in_seqs = x
-
-            # `inner_scan_inputs` is a list with the following composite form:
-            # inner_in_seqs
-            # + sum(inner_in_mit_mot, [])
-            # + sum(inner_in_mit_sot, [])
-            # + inner_in_sit_sot
-            # + inner_in_shared
-            # + inner_in_non_seqs
-            inner_in_mit_sot_flatten = []
-            for array, index in zip(inner_in_mit_sot, scan_args.mit_sot_in_slices):
-                inner_in_mit_sot_flatten.extend(array[jnp.array(index)])
-
-            inner_scan_inputs = sum(
-                [
-                    inner_in_seqs,
-                    inner_in_mit_mot,
-                    inner_in_mit_sot_flatten,
-                    inner_in_sit_sot,
-                    inner_in_shared,
-                    inner_in_non_seqs,
-                ],
-                [],
-            )
+            # `x` contains the inner sequences
+            inner_seqs = x
+
+            mit_sot_flatten = []
+            for array, index in zip(inner_mit_sot, op.info.mit_sot_in_slices):
+                mit_sot_flatten.extend(array[jnp.array(index)])
+
+            inner_scan_inputs = [
+                *inner_seqs,
+                *mit_sot_flatten,
+                *inner_sit_sot,
+                *inner_shared,
+                *inner_non_seqs,
+            ]
 
             return inner_scan_inputs
 
-        def inner_scan_outs_to_jax_outs(
-            op,
+        def inner_func_outs_to_jax_outs(
             old_carry,
             inner_scan_outs,
         ):
+            """Convert inner_scan_func outputs into format expected by JAX scan.
+
+            old_carry + (mit_sot_outs, sit_sot_outs, nit_sot_outs, shared_outs) -> (new_carry, ys)
+            """
             (
-                inner_in_mit_mot,
-                inner_in_mit_sot,
-                inner_in_sit_sot,
-                inner_in_shared,
-                inner_in_non_seqs,
+                inner_mit_sot,
+                inner_sit_sot,
+                inner_shared,
+                inner_non_seqs,
             ) = old_carry
 
-            def update_mit_sot(mit_sot, new_val):
-                return jnp.concatenate([mit_sot[1:], new_val[None, ...]], axis=0)
+            inner_mit_sot_outs = op.inner_mitsot_outs(inner_scan_outs)
+            inner_sit_sot_outs = op.inner_sitsot_outs(inner_scan_outs)
+            inner_nit_sot_outs = op.inner_nitsot_outs(inner_scan_outs)
+            inner_shared_outs = op.inner_shared_outs(inner_scan_outs)
+
+            # Replace the oldest mit_sot tap by the newest value
+            inner_mit_sot_new = []
+            if inner_mit_sot:
+                inner_mit_sot_new = [
+                    jnp.concatenate([old_mit_sot[1:], new_val[None, ...]], axis=0)
+                    for old_mit_sot, new_val in zip(
+                        inner_mit_sot,
+                        inner_mit_sot_outs,
+                    )
+                ]
```

Comment:
I thought the `[old_mit_sot[1:]]` would fail, but slicing works as expected. Still I feel this is a bit more readable?

Reply:
Regardless of whether or not it would fail, it is actually never reached because the list comprehension is empty.
I prefer the streamlined version without the condition, but it is your call.

Reply:
Removed